src/core/experimental/refcount.d

```diff
@@ -0,0 +1,548 @@
+         to this `const pure` function.
+    */
+    @nogc nothrow pure @safe scope
+    private void* addRef(this Q)() const
+    {
+        assert(isInitialized(), "[__RefCount.addRef] __RefCount is uninitialized");
+        cast(void) rcOp!(Q, "+=")(1);
+        return null;
+    }
+
+    /*
+    Decrease the reference count. If this was the last reference, `free` the
+    support. This asserts that `__RefCount` is initialized.
+
+    Returns:
+         This returns a `void*` so the compiler won't optimize away the call
+         to this `const pure` function.
+    */
+    @nogc nothrow pure @trusted scope
+    private void* delRef(this Q)() const
+    {
+        assert(isInitialized(), "[__RefCount.delRef] __RefCount is uninitialized");
+        /*
+         * This is an optimization. Most likely, most of the time, the refcount
+         * is `1`, so we don't want to make more ops to update that value only
+         * to free it afterwards. This is why we first check if the value is `1`.
+         * The `or` part decrements and checks for `0` in the case that we had
+         * two threads that both questioned at the same time if the value was `1`;
+         * they both will decrement, but only one will have the final value `0`.
+         */
+        if (rcOp!(Q, "==")(1) || (rcOp!(Q, "-=")(1) == 0))
+        {
+            return deallocate();
+        }
+        return null;
+    }
+
+    /*
+    `free` the support.
+
+    Returns:
+         This returns a $(B void*) so the compiler won't optimize away the call
+         to this `const pure` function.
+    */
+    @nogc nothrow pure @system scope
+    private void* deallocate(this Q)() const
+    {
+        return pureDeallocate((cast(CounterType*) rc)[0 .. 1]);
+    }
+
+    /**
+    Destruct the `__RefCount`. If it's initialized, decrement the refcount.
+    */
+    @nogc nothrow pure @trusted scope
+    ~this()
+    {
+        if (isInitialized())
+        {
+            delRef();
+        }
+    }
+
+    /**
+    Return a boolean value denoting if this is the only reference to this object.
+
+    Returns:
+         `true` if this reference count is unique; `false` if this `__RefCount`
+         object is uninitialized or there are multiple references to it.
+
+    Complexity:
+         $(BIGOH 1).
+    */
+    pure nothrow @safe @nogc scope
+    bool isUnique(this Q)() const
+    {
+        return isInitialized() && (!!rcOp!(Q, "==")(1));
+    }
+
+    /**
+    Return a boolean value denoting if this `__RefCount` object is initialized.
+
+    Returns:
+         `true` if initialized; `false` otherwise
+
+    Complexity:
+         $(BIGOH 1).
+    */
+    pure nothrow @safe @nogc scope
+    bool isInitialized(this Q)() const
+    {
+        return rc !is null;
+    }
+
+    version (CoreUnittest)
+    {
+        pure nothrow @nogc @trusted scope
+        private bool isValueEq(this Q)(uint val) const
+        {
+            return *getUnsafeValue == val;
+        }
+    }
```

Comment:
Why do we need this? Using `version(unittest)` directives has bitten us a lot in the past as all tests passed, but the end-users got a different behavior -> I recommend moving this into the actual `unittest` block.
One can access private members in the same file.